File: mobile/lib/fxrates.js
// fxrates.js: the pure, testable core of live exchange rates.
//
// Salapify stays offline first. This layer only makes the app NICER when it has
// internet: it fetches today's public rates so the log sheet can pre fill the
// rate for a foreign expense. It never sends any user or financial data, it only
// downloads a table of public market rates. When the phone is offline, the fetch
// fails, or a currency is not covered, the user types the rate by hand exactly as
// before, so nothing here is ever load bearing for correctness.
//
// The network fetch and cache live in hooks/useFxRates.js (they touch the device);
// everything here is pure so it can be unit tested.

// Free, no API key. The response shape is documented and stable:
//   { result: 'success', base_code: 'PHP', time_last_update_unix: 1710000000,
//     rates: { USD: 0.0176, JPY: 2.62, ... } }  // rates are UNITS PER 1 base
export const FX_PROVIDER = 'open.er-api.com';
export const FX_ENDPOINT = (base) =>
  `https://open.er-api.com/v6/latest/${encodeURIComponent(String(base || 'PHP'))}`;

// Refetch at most twice a day. Rates barely move day to day for budgeting, and a
// stale-by-a-few-hours rate the user can see and override is fine.
export const FX_MAX_AGE_MS = 12 * 60 * 60 * 1000;

// Turn the provider response into { base, rates, fetchedAt } or null. Anything
// unexpected returns null so the caller falls back to a typed rate.
export function parseRatesResponse(json) {
  if (!json || json.result !== 'success' || !json.rates || typeof json.rates !== 'object') return null;
  const base = typeof json.base_code === 'string' ? json.base_code : null;
  if (!base) return null;
  const fetchedAt = Number(json.time_last_update_unix);
  return {
    base,
    rates: json.rates,
    fetchedAt: Number.isFinite(fetchedAt) && fetchedAt > 0 ? fetchedAt * 1000 : null,
  };
}

// base currency per 1 unit of `code`. The provider gives units-per-base, so we
// invert. Returns null when the code is missing or the rate is not a positive
// number, so a bad or absent entry never yields a wrong figure.
export function basePerUnit(rates, code) {
  const perBase = rates ? Number(rates[code]) : NaN;
  if (!Number.isFinite(perBase) || perBase <= 0) return null;
  return 1 / perBase;
}

// Round a rate to four significant figures so the pre filled value is tidy for
// both strong (56.34) and weak (0.002315) currencies without losing accuracy.
export function roundRate(r) {
  const n = Number(r);
  if (!Number.isFinite(n) || n <= 0) return null;
  return Number(n.toPrecision(4));
}

// Is a cached table still fresh enough to skip a refetch?
export function isFresh(fetchedAt, nowMs) {
  if (!fetchedAt) return false;
  return nowMs - fetchedAt < FX_MAX_AGE_MS;
}

File: flutter/lib/money/fxrates.dart
// The pure, testable core of live exchange rates, ported 1:1 from
// mobile/lib/fxrates.js. Salapify stays offline first: this only makes the app
// nicer when it has internet by fetching today's public rates. It never sends
// any user or financial data, only downloads a table of public market rates.
// Offline, on a failed fetch, or an uncovered currency, the user types the rate
// by hand, so nothing here is ever load bearing for correctness. The network
// fetch and cache live in the fx service (it touches the device); everything
// here is pure and golden verified against the RN app.

/// Free, no API key. Rates are UNITS PER 1 base.
const String fxProvider = 'open.er-api.com';
String fxEndpoint(String? base) =>
    'https://open.er-api.com/v6/latest/${Uri.encodeComponent(base == null || base.isEmpty ? 'PHP' : base)}';

/// Refetch at most twice a day; rates barely move for budgeting.
const int fxMaxAgeMs = 12 * 60 * 60 * 1000;

double? _rate(dynamic rates, dynamic code) {
  if (rates is! Map) return null;
  final v = rates[code];
  final n = v is num ? v.toDouble() : double.tryParse('${v ?? ''}');
  return (n != null && n.isFinite) ? n : null;
}

/// Turn the provider response into { base, rates, fetchedAt } or null. Anything
/// unexpected returns null so the caller falls back to a typed rate.
Map<String, dynamic>? parseRatesResponse(dynamic json) {
  if (json is! Map ||
      json['result'] != 'success' ||
      json['rates'] is! Map) {
    return null;
  }
  final base = json['base_code'] is String ? json['base_code'] as String : null;
  if (base == null) return null;
  final raw = json['time_last_update_unix'];
  final fetched = raw is num ? raw.toDouble() : double.tryParse('${raw ?? ''}');
  return {
    'base': base,
    'rates': json['rates'],
    'fetchedAt':
        (fetched != null && fetched.isFinite && fetched > 0) ? fetched * 1000 : null,
  };
}

/// base currency per 1 unit of `code`. The provider gives units-per-base, so we
/// invert. Null when the code is missing or the rate is not a positive number.
double? basePerUnit(dynamic rates, dynamic code) {
  final perBase = _rate(rates, code);
  if (perBase == null || perBase <= 0) return null;
  return 1 / perBase;
}

/// The cross rate to convert 1 unit of `from` into `to` using one units-per-base
/// table. rates[to] / rates[from] cancels the base out. Null when either
/// currency is missing or non positive, so a gap never yields a wrong figure.
double? crossRate(dynamic rates, dynamic from, dynamic to) {
  final f = _rate(rates, from);
  final t = _rate(rates, to);
  if (f == null || f <= 0 || t == null || t <= 0) return null;
  return t / f;
}

/// Round a rate to four significant figures so the pre filled value is tidy for
/// both strong (56.34) and weak (0.002315) currencies. Mirrors JS
/// Number(n.toPrecision(4)); null on a non positive or non finite input.
double? roundRate(dynamic r) {
  final n = r is num ? r.toDouble() : double.tryParse('${r ?? ''}');
  if (n == null || !n.isFinite || n <= 0) return null;
  return double.parse(n.toStringAsPrecision(4));
}

/// Is a cached table still fresh enough to skip a refetch? A null or zero stamp
/// is never fresh (JS treats both as falsy).
bool isFresh(dynamic fetchedAt, num nowMs) {
  final f = fetchedAt is num ? fetchedAt.toDouble() : double.tryParse('${fetchedAt ?? ''}');
  if (f == null || f == 0) return false;
  return nowMs - f < fxMaxAgeMs;
}
